mm2src/coins/nft_storage/sql_storage.rs

```diff
@@ -0,0 +1,885 @@
+    let mut sql_builder = SqlQuery::select_from(conn, table_name)?;
+    if let Some(filters) = filters {
+        if filters.send && !filters.receive {
+            sql_builder.sql_builder().and_where_eq("status", "'Send'");
+        } else if filters.receive && !filters.send {
+            sql_builder.sql_builder().and_where_eq("status", "'Receive'");
+        }
+
+        if filters.from_date.is_some() && filters.to_date.is_some() {
+            sql_builder.sql_builder().and_where(format!(
+                "block_timestamp BETWEEN '{}' and '{}'",
+                filters.from_date.unwrap(),
+                filters.to_date.unwrap()
+            ));
+        } else if filters.from_date.is_some() {
+            sql_builder
+                .sql_builder()
+                .and_where(format!("block_timestamp >= '{}'", filters.from_date.unwrap()));
+        } else if filters.to_date.is_some() {
+            sql_builder
+                .sql_builder()
+                .and_where(format!("block_timestamp <= '{}'", filters.to_date.unwrap()));
+        }
+    }
+    drop_mutability!(sql_builder);
+    Ok(sql_builder)
+}
+
+fn nft_table_builder_preimage<'a>(conn: &'a Connection, table_name: &'a str) -> MmResult<SqlQuery<'a>, SqlError> {
+    let sql_builder = SqlQuery::select_from(conn, table_name)?;
+    Ok(sql_builder)
+}
+
+fn finalize_nft_list_sql_builder(sql_builder: &mut SqlQuery, offset: usize, limit: usize) -> MmResult<(), SqlError> {
+    sql_builder.field("nft_list.details_json")?.offset(offset).limit(limit);
+    Ok(())
+}
+
+fn finalize_nft_history_sql_builder(sql_builder: &mut SqlQuery, offset: usize, limit: usize) -> MmResult<(), SqlError> {
+    sql_builder
+        .field("nft_history.details_json")?
+        .offset(offset)
+        .limit(limit);
+    Ok(())
+}
+
+fn nft_from_row(row: &Row<'_>) -> Result<Nft, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn tx_history_from_row(row: &Row<'_>) -> Result<NftTransferHistory, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn token_address_id_from_row(row: &Row<'_>) -> Result<NftTokenAddrId, SqlError> {
+    let token_address: String = row.get("token_address")?;
+    let token_id_str: String = row.get("token_id")?;
+    let token_id = BigDecimal::from_str(&token_id_str).map_err(|_| SqlError::from(FromSqlError::InvalidType))?;
+    Ok(NftTokenAddrId {
+        token_address,
+        token_id,
+    })
+}
+
+fn insert_nft_in_list_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_list_table_name(chain);
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            token_address, token_id, chain, amount, block_number, contract_type, details_json
```

Comment:
Interested to know why `INSERT OR IGNORE` is used here instead of `INSERT OR REPLACE`. Probably it has lower cost to ignore if the entry is in the table but then I wonder why we have a value that can change such as amount as part of the table columns.

Reply:
I intentionally separated INSERT ans UPDATE methods so as not to confuse the logic.
and yeah, you are right its better to return error if we try to insert already existed entries.
I added INSERT INTO in `insert_nft_in_list_sql` and `insert_tx_in_history_sql`.

> I wonder why we have a value that can change such as amount as part of the table columns

Weeell, I decided to leave it in the case if we have filters for amount in the future. Even if I remove it from table, I still have to call update method to change the amount in details json field.